File: anagram/anagram_solver.py
#!/usr/bin/env python3
from collections import defaultdict
import sys

def load_wordlist(filename="wordlist.txt"):
    anagram_map = defaultdict(set)
    try:
        with open(filename, "r", encoding='utf-8') as f:
            for line in f:
                word = line.strip().upper()
                key = ''.join(sorted(word))
                anagram_map[key].add(word)
    except FileNotFoundError:
        print("Erreur : impossible d'ouvrir wordlist.txt")
        sys.exit(1)
    return anagram_map

def doit(letters, anagram_map):
    letters = letters.upper()
    available = sorted(letters)
    n = min(len(available), 15)  # Limite à 15 lettres pour éviter explosion
    found_words = set()

    for i in range(1, 1 << n):  # 2^n combinaisons
        candidate = []
        for j in range(n):
            if i & (1 << j):
                candidate.append(available[j])
        sorted_candidate = ''.join(candidate)
        if sorted_candidate in anagram_map:
            found_words.update(anagram_map[sorted_candidate])

    # Grouper par longueur
    grouped = defaultdict(list)
    for word in found_words:
        grouped[len(word)].append(word)

    # Afficher
    print(f"-------------------------")
    print(f"Anagrammes de [{letters} - {len(letters)}]")
    print(f"-------------------------")

    for length in sorted(grouped):
        words = ', '.join(sorted(grouped[length]))
        print(f"[{length}] {words}")

def main():
    anagram_map = load_wordlist()

    if len(sys.argv) > 1:
        doit(sys.argv[1], anagram_map)
    else:
        while True:
            try:
                letters = input("\nLettres: ").strip()
                if not letters:
                    break
                doit(letters, anagram_map)
            except (EOFError, KeyboardInterrupt):
                break

if __name__ == "__main__":
    main()

File: anagram/AnagramSolver.java

import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

public class AnagramSolver {

    private static Map<String, Set<String>> anagramMap = new HashMap<>();

    public static void main(String[] args) throws IOException {
        loadWordlist("wordlist.txt");

        if (args.length > 0) {
            doit(args[0]);
        } else {
            Scanner scanner = new Scanner(System.in);
            while (true) {
                System.out.print("\nLettres: ");
                System.out.flush();

                String input = scanner.nextLine().trim();
                if (input.isEmpty()) break;

                doit(input);
            }
        }
    }

    private static void loadWordlist(String filename) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.trim().toUpperCase();
                char[] chars = line.toCharArray();
                Arrays.sort(chars);
                String sorted = new String(chars);

                anagramMap.computeIfAbsent(sorted, k -> new HashSet<>()).add(line);
            }
        }
    }

    private static void doit(String letters) {
        letters = letters.toUpperCase();
        char[] available = letters.toCharArray();
        Arrays.sort(available);

        Set<String> foundWords = new TreeSet<>(); // Pour tri automatique

        int n = Math.min(available.length, 15); // Limite à 15 lettres
        int mask = 1 << n;

        for (int i = 1; i < mask; i++) {
            StringBuilder candidate = new StringBuilder();
            for (int j = 0; j < n; j++) {
                if ((i & (1 << j)) != 0) {
                    candidate.append(available[j]);
                }
            }

            String sortedCandidate = candidate.toString();
            Set<String> matches = anagramMap.get(sortedCandidate);
            if (matches != null) {
                foundWords.addAll(matches);
            }
        }

        // Grouper par longueur
        Map<Integer, List<String>> groupedByLength = foundWords.stream()
            .collect(Collectors.groupingBy(String::length));

        System.out.println("-------------------------");
        System.out.println("Anagrammes de [" + letters + " - " + letters.length() + "]");
        System.out.println("-------------------------");

        groupedByLength.entrySet().stream()
            .sorted(Map.Entry.comparingByKey())
            .forEach(entry -> {
                System.out.println("[" + entry.getKey() + "] " + String.join(", ", entry.getValue()));
            });
    }
}